GSUP: Send InsertSubscriberData on subscriber change

Takuto88 opened the pull request:
When the subscriber is updated, e.g. the APNs or the MSISDNs change, we must send an update to the relevant network elements like MSC or SGSN. This PR implements this by using Redis messaging to communicate to the GSUP daemon that the Subscriber has changed, which will then  tigger an ISD to each relevant peer.

I have tried to not repeat myself and share code between the similar UpdateLocation procedure. Since both procedures are somewhat transactional, I've formalized the transaction procedure into an `AbstractTransaction` that is extended by both the existing `ULRTransaction` and the new `ISDTransaction`.

The code works and is functional but I am looking forward to feedback on this as there may be some areas where it could improve. For example, the creation of the InsertSubscriberData Request lives in the abstract transaction out of convenience and that might not be the best place for it.

This PR also will repair the APN data for the InsertSubscriberData Request - Those were previously not transmitted at all. I am not sure if this is correct. Thus I've attached some PCAPs for easy review:

[pcaps.zip](https://github.com/user-attachments/files/23637753/pcaps.zip)

osmith42 commented on lib/gsup/server.py:
```diff
@@ -216,3 +221,22 @@ async def __handle_gsup(self, peer: IPAPeer, payload_length: int):
+
+            except Exception as e:
+                await self.logger.logAsync(service='GSUP', level='ERROR',
+                                           message=f"Error processing subscriber update: {e}")
```
How about logging a full traceback here too on error?

```suggestion
                                           message=f"Error processing subscriber update: {traceback.format_exc()}")
```

Otherwise it is really hard to understand some errors, e.g. I just got this because I didn't have redis running while testing the PR:

```
[11/20/2025 10:12:46] [ERROR] Error processing subscriber update: not enough values to unpack (expected 2, got 0)
```

With the traceback:

```
[11/20/2025 10:14:28] [ERROR] Error processing subscriber update: Traceback (most recent call last):
  File "/home/user/.cache/osmo-ttcn3-testenv/host/install/venv/lib/python3.12/site-packages/pyhss/lib/gsup/server.py", line 232, in _listen_for_subscriber_updates
    _, message_data = await asyncio.to_thread(self.redis_messaging.awaitMessage, 'subscriber_update')
    ^^^^^^^^^^^^^^^
ValueError: not enough values to unpack (expected 2, got 0)
```

osmith42 commented on lib/gsup/controller/abstract_transaction.py:
```diff
@@ -0,0 +1,65 @@
+
+        if cn_domain == 'ps':
+            for index, apn in enumerate(subscriber_info.apns):
+                request_builder.with_pdp_info_ie(index, apn['ip_version'], apn['name'])
```
Actually instead of the suggestion above: it would probably make more sense to use the apn_id from the database:

https://github.com/nickvsnetworking/pyhss/blob/7afa965e3b2f2827e9cbd43b9f60586511377140/lib/database.py#L31

Takuto88 replied:
Done [e2b3816](https://github.com/nickvsnetworking/pyhss/pull/277/commits/e2b3816bab094cf056e1e620108634861a455484)

Takuto88 commented on lib/gsup/controller/abstract_transaction.py:
```diff
@@ -0,0 +1,42 @@
+class AbstractTransaction(ABC):
+
+    def __init__(self):
+        self._started_at = datetime.now()
```
Done [3e5e52a](https://github.com/nickvsnetworking/pyhss/pull/277/commits/3e5e52a8491ea306d3bee6a99bd2633aeae17395)

Takuto88 commented:
I've addressed all the Feedback in this PR. While testing, I also found another small bug that added the IMSI IE twice to the ULA. While this is technically out of scope, I've snuck this fix into this PR as well: [ae3cc4b](https://github.com/nickvsnetworking/pyhss/pull/277/commits/ae3cc4b7f7de32cd07dda02920ff209d9993674e)

osmith42 commented on lib/gsup/controller/abstract_transaction.py:
```diff
@@ -0,0 +1,65 @@
+                           .with_msg_type(MsgType.INSERT_DATA_REQUEST)
+                           .with_ie('imsi', subscriber_info.imsi)
+                           .with_msisdn_ie(subscriber_info.msisdn)
+                           )
```
Otherwise LGTM, feel free to merge after this is addressed.
